alchemiscale/tests/integration/storage/test_statestore.py

```diff
@@ -2225,6 +2294,153 @@ def test_get_task_restart_patterns(self, n4js, network_tyk2, scope_test):
+            assert len(tasks_actioned_by_all_taskhubs) == 4
+
+            # we're going to just pass the first 2 and fail the second 2
+            tasks_to_complete = tasks_actioned_by_all_taskhubs[:2]
+            tasks_to_fail = tasks_actioned_by_all_taskhubs[2:]
+
+            complete_tasks(n4js, tasks_to_complete)
+
+            records = n4js.execute_query(
+                """
+                UNWIND $task_scoped_keys as task_scoped_key
+                MATCH (task:Task {_scoped_key: task_scoped_key})-[:RESULTS_IN]->(:ProtocolDAGResultRef)
+                RETURN count(task) as task_count
+            """,
+                task_scoped_keys=list(map(str, tasks_to_complete)),
+            ).records
+
+            assert records[0]["task_count"] == 2
+
+            # test the behavior of the compute API
+            for i, task in enumerate(tasks_to_fail):
+                error_messages = [
+                    f"Error message {repeat}, round {i}" for repeat in range(3)
+                ]
+
+                fail_task(
+                    n4js,
+                    task,
+                    resolve=False,
+                    error_messages=error_messages,
+                )
+
+                n4js.resolve_task_restarts(all_tasks)
+
+            # both tasks should have the waiting status and the APPLIES
+            # relationship num_retries should have incremented by 1
+            query = """
+            UNWIND $task_scoped_keys as task_scoped_key
+            MATCH (task:Task {`_scoped_key`: task_scoped_key, status: $waiting})<-[:APPLIES {num_retries: 1}]-(:TaskRestartPattern {max_retries: 2})
+            RETURN count(DISTINCT task) as renewed_waiting_tasks
+            """
+
+            renewed_waiting = n4js.execute_query(
+                query,
+                task_scoped_keys=list(map(str, tasks_to_fail)),
+                waiting=TaskStatusEnum.waiting.value,
+            ).records[0]["renewed_waiting_tasks"]
+
+            assert renewed_waiting == 2
+
+            # we want the resolve restarts to cancel a task.
+            # deconstruct the tasks to fail, where the first
+            # one will be cancelled and the second will once again be continued
+            # but with an additional traceback
+            task_to_cancel, task_to_wait = tasks_to_fail
+
+            for _ in range(2):
+                error_messages = [
+                    f"Error message {repeat}, round {i}" for repeat in range(3)
+                ]
+
+                fail_task(
+                    n4js,
+                    task_to_cancel,
+                    resolve=False,
+                    error_messages=error_messages,
+                )
+
+                n4js.resolve_task_restarts(tasks_to_fail)
+
+            assert tasks_are_not_actioned_on_taskhub(
+                n4js,
+                [task_to_cancel],
+                taskhub_scoped_key_with_policy,
+            )
+
+            assert tasks_are_errored(n4js, [task_to_cancel])
+
```

Comment:
Is there more you intend to add to this test? I see in the comment above that the second one will be continued but with an additional traceback? Are you going to try matching to a different restart pattern for that one?